rom 427b560f8d0808d2a5beb76966b3c6c575d2fa75 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Daniel=20Vr=C3=A1til?= <dvratil@kde.org>
Date: Wed, 19 Apr 2017 17:46:06 +0200
Subject: [IMAP] Fix missing connect() when restarting IDLE job

Restarted IDLE job did not listen for flags changes.
---
 resources/imap/imapidlemanager.cpp  | 10 ++++++----
 resources/imap/imapidlemanager.h    |  2 ++
 resources/imap/imapresourcebase.cpp |  2 ++
 3 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/resources/imap/imapidlemanager.cpp b/resources/imap/imapidlemanager.cpp
index 76d48c2..171face 100644
--- a/resources/imap/imapidlemanager.cpp
+++ b/resources/imap/imapidlemanager.cpp
@@ -97,6 +97,11 @@ void ImapIdleManager::onSessionRequestDone(qint64 requestId, KIMAP::Session *ses
     connect(m_pool, &SessionPool::connectionLost, this, &ImapIdleManager::onConnectionLost);
     connect(m_pool, &SessionPool::disconnectDone, this, &ImapIdleManager::onPoolDisconnect);
 
+    startIdle();
+}
+
+void ImapIdleManager::startIdle()
+{
     KIMAP::SelectJob *select = new KIMAP::SelectJob(m_session);
     select->setMailBox(m_state->mailBoxForCollection(m_state->collection()));
     connect(select, &KIMAP::SelectJob::result, this, &ImapIdleManager::onSelectDone);
@@ -142,10 +147,7 @@ void ImapIdleManager::onIdleStopped()
     m_idle = nullptr;
     if (m_session) {
         qCDebug(IMAPRESOURCE_LOG) << "Restarting the IDLE session!";
-        m_idle = new KIMAP::IdleJob(m_session);
-        connect(m_idle.data(), &KIMAP::IdleJob::mailBoxStats, this, &ImapIdleManager::onStatsReceived);
-        connect(m_idle.data(), &KIMAP::IdleJob::result, this, &ImapIdleManager::onIdleStopped);
-        m_idle->start();
+        startIdle();
     }
 }
 
diff --git a/resources/imap/imapidlemanager.h b/resources/imap/imapidlemanager.h
index 4a2f514..df7f03d 100644
--- a/resources/imap/imapidlemanager.h
+++ b/resources/imap/imapidlemanager.h
@@ -69,6 +69,8 @@ private Q_SLOTS:
     void reconnect();
 
 private:
+    void startIdle();
+
     qint64 m_sessionRequestId;
     SessionPool *m_pool;
     KIMAP::Session *m_session;
diff --git a/resources/imap/imapresourcebase.cpp b/resources/imap/imapresourcebase.cpp
index 10a9bad..5ff370b 100644
--- a/resources/imap/imapresourcebase.cpp
+++ b/resources/imap/imapresourcebase.cpp
@@ -632,6 +632,8 @@ void ImapResourceBase::onIdleCollectionFetchDone(KJob *job)
     if (!fetch->collections().isEmpty()) {
         delete m_idle;
         m_idle = new ImapIdleManager(createResourceState(TaskArguments(fetch->collections().at(0))), m_pool, this);
+    } else {
+        qCWarning(IMAPRESOURCE_LOG) << "Failed to retrieve IDLE collection: no such collection";
     }
 }
 
-- 
cgit v0.11.2

